server/service/hosts.go

```diff
@@ -186,18 +192,36 @@ func deleteHostsEndpoint(ctx context.Context, request interface{}, svc fleet.Ser
 		StatusFilter: req.Filters.Status,
 		TeamFilter:   req.Filters.TeamID,
 	}
-	err := svc.DeleteHosts(ctx, req.IDs, listOpt, req.Filters.LabelID)
-	if err != nil {
-		return deleteHostsResponse{Err: err}, nil
+
+	// Since bulk deletes can take a long time, after DeleteHostsTimeout, we will return a 202 (Accepted) status code
+	// and allow the delete operation to proceed.
+	var err error
+	deleteDone := make(chan bool, 1)
+	ctx = context.WithoutCancel(ctx) // to make sure DB operations don't get killed after we return a 202
+	go func() {
+		err = svc.DeleteHosts(ctx, req.IDs, listOpt, req.Filters.LabelID)
+		deleteDone <- true
+	}()
+	select {
+	case <-deleteDone:
+		if err != nil {
+			return deleteHostsResponse{Err: err}, nil
+		}
+		return deleteHostsResponse{StatusCode: http.StatusOK}, nil
+	case <-time.After(DeleteHostsTimeout):
+		return deleteHostsResponse{StatusCode: http.StatusAccepted}, nil
 	}
-	return deleteHostsResponse{}, nil
 }
 
 func (svc *Service) DeleteHosts(ctx context.Context, ids []uint, opts fleet.HostListOptions, lid *uint) error {
 	if err := svc.authz.Authorize(ctx, &fleet.Host{}, fleet.ActionList); err != nil {
 		return err
 	}
 
+	if len(ids) == 0 && lid == nil && opts.Empty() {
+		return &fleet.BadRequestError{Message: "list of ids or filters must be specified"}
+	}
+
```

Comment:
This was a bug I found -- if a user doesn't specify or misspell a parameter, all their hosts get deleted.

Reply:
Great catch!

Reply:
It seems like the issue stems in large part from the `hostIDsAndNamesFromFilters` service [method](https://github.com/fleetdm/fleet/blob/76059564a922e1bcf29a89499719b80e6d9e710c/server/service/hosts.go#L995). Should we also look into how we might be misusing that in other places for write operations (e.g., do we allow transfer hosts from filters)? Any ideas on ways we can improve that method to make it less of a trap for the unwary in the future?